### DISCUSSION TITLE
dynamic query and mutation

### DISCUSSION THREAD

**Discussion by arashi-dev** [trpc/trpc#3592] [Q&A]
Hello. I have an `edit form` component like this:

```tsx
    const router = useRouter()
    const { t } = useTranslation(["admin/common"])
    const productId = useMemo(() => router.isReady ? (Array.isArray(router.query.id) ? router.query.id[0] : router.query.id) : undefined, [router.isReady, router.query.id])

    const { data, status } = trpc.admin.products.findOne.useQuery({
        id: +productId!,
    }, {
        enabled: !!productId
    })

    const utils = trpc.useContext()
    const mutate = trpc.admin.products.update.useMutation({
        onSuccess() {
            showNotification({
                message: t("admin/common:messages.success.edit"),
                color: "green"
            })
            utils.admin.products.getAll.invalidate()
        },
        onError() {
            showNotification({
                message: t("admin/common:messages.error.unexpected"),
                color: "red"
            })
        }
    })

    const form = useForm({
        initialValues: data!,
        validateInputOnChange: true,
        validate: zodResolver(productSchema.update),
    })

    ...
```

this code is going to be repeated for different models (Product, User, Color, Category, etc.). so, I am looking for a way to make it DRY and reusable.
the only difference between them is the model name in trpc ` trpc.admin[THE_MODEL_NAME]`. the rest are just repeated and the same.

how can I make it reusable? I thought maybe I can create a Component for it and pass `fetchData`, `updateData` props to handle the query and mutations. but they are harder to manage as I can't track the status and abort on unmount

**Comment by timcole** [comment]
Heyo! @jack-douglas and I are running into a similar issue. We're working with a bunch of different social media websites that have slightly different types to each other. The data we need to access is all the same though some things like name and id are guaranteed whereas something like a verification status may be missing from another type. Because the data we need to access is all the same, we want to do something like this so we don't have to duplicate the same code for every platform. Please note this functionality worked perfectly in v9.

Here's a small codesandbox showing the issue in a similar way we're trying to use it. Our actual use-case involves a lot more platforms than just two used in the example but the ts errors in this still apply. https://codesandbox.io/p/sandbox/practical-goldberg-or2fin?file=%2Fcomponents%2Fuser.tsx%3A9%2C21

If maybe @KATT or someone could take a look and give us some pointers or help us with a solution that would be amazing!
